### PR TITLE
add etcd registry and resolver implementation

### PR DESCRIPTION
**Pull request by lizhemingi**
Related [#155](https://github.com/cloudwego/kitex/issues/155)

### CODE REVIEW

**Comment by YangruiEmma** [comment]
Thank you very much! Sorry, it has been a long time. We will review it.

File: common.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package etcd
+
+const (
+	etcdPrefix = "kitex/registry-etcd"
+)
+
+func serviceKeyPrefix(servieName string) string {
```
Comment (ppzqh) [inline common.go]: servieName -> serviceName

File: etcd_resolver.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+		info serviceInfo
+		eps  []discovery.Instance
+	)
+	for _, kv := range resp.Kvs {
+		err := json.Unmarshal(kv.Value, &info)
+		if err != nil {
+			klog.Warnf("fail to unmarshal with err: %v, ignore key: %v", err, string(kv.Key))
+			continue
+		}
+		weight := info.Weight
```
Comment (ppzqh) [inline etcd_resolver.go:80]: Add some checks about the info? like info.Address != ""

File: etcd_registry.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+		return fmt.Errorf("missing service name in Deregister")
+	}
+	if err := e.deregist(info); err != nil {
+		return err
+	}
+	e.meta.cancel()
+	return nil
+}
+
+func (e *etcdRegistry) regist(info *registry.Info, leaseID clientv3.LeaseID) error {
```
Comment (ppzqh) [inline etcd_registry.go]: regist -> register

File: etcd_registry.go [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+	if err != nil {
+		return err
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*3)
+	defer cancel()
+	_, err = e.etcdClient.Put(ctx, serviceKey(info.ServiceName, info.Addr.String()), string(val), clientv3.WithLease(leaseID))
+	return err
+}
+
+func (e *etcdRegistry) deregist(info *registry.Info) error {
```
Comment (ppzqh) [inline etcd_registry.go]: deregist -> deregister

File: etcd_registry.go [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+	defaultTTL = 60
+)
+
+type etcdRegistry struct {
+	etcdClient *clientv3.Client
+	leaseTTL   int64
+	meta       *registMeta
+}
+
+type registMeta struct {
```
Comment (ppzqh) [inline etcd_registry.go]: registMeta -> registerMeta

File: common.go [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+	return etcdPrefix + "/" + servieName
+}
+
+// serviceKey generates the key used to stored in etcd.
+func serviceKey(serviceName, addr string) string {
+	return serviceKeyPrefix(serviceName) + "/" + addr
+}
+
+// serviceInfo used to stored service basic info in etcd.
+type serviceInfo struct {
```
Comment (ppzqh) [inline common.go]: I think `instanceInfo` is a better variable name.

File: etcd_registry.go [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+	leaseID clientv3.LeaseID
+	ctx     context.Context
+	cancel  context.CancelFunc
+}
+
+// NewEtcdRegistry creates a etcd based registry.
+func NewEtcdRegistry(endpoints []string) (registry.Registry, error) {
```
Comment (ppzqh) [inline etcd_registry.go:49]: Add functions with authentication for both Registry and Resolver.
like func NewEtcdRegistryWithAuth(endpoints []string, user, password string)

set the `Username` and `Password` in `clientv3.Config`.

**Comment by lizhemingi** [comment]
@YangruiEmma A kindly ping in case you missed this pr. :)

**Review by ppzqh — COMMENTED** [review]
please check the reviews.

**Comment by ppzqh** [comment]
> @YangruiEmma A kindly ping in case you missed this pr. :)

@duduainankai Sorry for the delay.
Please check the reviews above and leave comments if there is any other advice. Thanks!

**Comment by lizhemingi** [comment]
@ppzqh PTAL.

File: etcd_registry.go [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+	return NewEtcdRegistryWithAuth(endpoints, "", "")
+}
+
+// NewEtcdRegistry creates a etcd based registry with given username and password.
```
Comment (ppzqh) [inline etcd_registry.go:53]: // NewEtcdRegistryWithAuth creates

**Comment by lizhemingi** [comment]
@ppzqh Hi, I saw the comment and just fix it already, but I don't know why it disappeared after I push the commit.  So I have nowhere to resolve .

File: README.md [hunk 9]
```diff
@@ -1 +1,59 @@
-# registry-etcd
\ No newline at end of file
+# registry-etcd
```
Comment (GuangmingLuo) [inline README.md]: Please add  (This is a community driven project) right after the title to show that this is a community driven project

**Comment by GuangmingLuo** [comment]
Moreover, you may add Compatibility section and Maintenance section in the README file.
Refer to :  https://github.com/kitex-contrib/registry-zookeeper

**Review by ppzqh — APPROVED** [review]

**Pull request merged**